include/core/mir/geometry/dimensions.h

```diff
@@ -29,180 +30,43 @@ namespace mir
-inline constexpr Height operator-(Height lhs, DeltaY rhs) { return Height(lhs.as_int() - rhs.as_int()); }
-inline Width& operator+=(Width& lhs, DeltaX rhs) { return lhs = lhs + rhs; }
-inline Height& operator+=(Height& lhs, DeltaY rhs) { return lhs = lhs + rhs; }
-inline Width& operator-=(Width& lhs, DeltaX rhs) { return lhs = lhs - rhs; }
-inline Height& operator-=(Height& lhs, DeltaY rhs) { return lhs = lhs - rhs; }
-
-// Adding Widths and Heights is fine
-inline constexpr Width operator+(Width lhs, Width rhs) { return Width(lhs.as_int() + rhs.as_int()); }
-inline constexpr Height operator+(Height lhs, Height rhs) { return Height(lhs.as_int() + rhs.as_int()); }
-inline Width& operator+=(Width& lhs, Width rhs) { return lhs = lhs + rhs; }
-inline Height& operator+=(Height& lhs, Height rhs) { return lhs = lhs + rhs; }
-
-// Subtracting coordinates is fine
-inline constexpr DeltaX operator-(X lhs, X rhs) { return DeltaX(lhs.as_int() - rhs.as_int()); }
-inline constexpr DeltaY operator-(Y lhs, Y rhs) { return DeltaY(lhs.as_int() - rhs.as_int()); }
-
-//Subtracting Width and Height is fine
-inline constexpr DeltaX operator-(Width lhs, Width rhs) { return DeltaX(lhs.as_int() - rhs.as_int()); }
-inline constexpr DeltaY operator-(Height lhs, Height rhs) { return DeltaY(lhs.as_int() - rhs.as_int()); }
-
-// Multiplying by a scalar value is fine
-template<typename Scalar>
-inline constexpr Width operator*(Scalar scale, Width const& w) { return Width{scale*w.as_int()}; }
-template<typename Scalar>
-inline constexpr Height operator*(Scalar scale, Height const& h) { return Height{scale*h.as_int()}; }
-template<typename Scalar>
-inline constexpr DeltaX operator*(Scalar scale, DeltaX const& dx) { return DeltaX{scale*dx.as_int()}; }
-template<typename Scalar>
-inline constexpr DeltaY operator*(Scalar scale, DeltaY const& dy) { return DeltaY{scale*dy.as_int()}; }
-template<typename Scalar>
-inline constexpr Width operator*(Width const& w, Scalar scale) { return scale*w; }
-template<typename Scalar>
-inline constexpr Height operator*(Height const& h, Scalar scale) { return scale*h; }
-template<typename Scalar>
-inline constexpr DeltaX operator*(DeltaX const& dx, Scalar scale) { return scale*dx; }
-template<typename Scalar>
-inline constexpr DeltaY operator*(DeltaY const& dy, Scalar scale) { return scale*dy; }
-
-// Dividing by a scaler value is fine
-template<typename Scalar>
-inline constexpr Width operator/(Width const& w, Scalar scale) { return Width{w.as_int() / scale}; }
-template<typename Scalar>
-inline constexpr Height operator/(Height const& h, Scalar scale) { return Height{h.as_int() / scale}; }
-template<typename Scalar>
-inline constexpr DeltaX operator/(DeltaX const& dx, Scalar scale) { return DeltaX{dx.as_int() / scale}; }
-template<typename Scalar>
-inline constexpr DeltaY operator/(DeltaY const& dy, Scalar scale) { return DeltaY{dy.as_int() / scale}; }
```

Comment:
The generic operators proposed do not replace these (they return the generic type not the derived type). Either the generic operators need to return the correct type, or these need to be reimplemented in terms of the generic operators.

Reply:
Yes, but they are implicitly and convertible to the the derived types at zero runtime cost, so the results turn into the derived types whenever we need them to. If we assign them to an `auto` variable they don't, but that's fine because the generic types can be used the same way. The only exception is when trying to use `std::min`/`max` with both generic and derived, but that only showed up a few places in the code base.